File: 3201-find-the-maximum-length-of-valid-subsequence-i/3201-find-the-maximum-length-of-valid-subsequence-i.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumLength = function(nums) {
    const parityList = nums.map(num => num % 2);
    let sumOfOnes = parityList.reduce((a, b) => a + b, 0);
    let alternatingLength = 1, previousParity = parityList[0];
    for (let currentParity of parityList.slice(1)) {
        if (previousParity !== currentParity) {
            alternatingLength += 1;
            previousParity = currentParity;
        }
    }
    return Math.max(sumOfOnes, parityList.length - sumOfOnes, alternatingLength);
};

File: 3201-find-the-maximum-length-of-valid-subsequence-i/3201-find-the-maximum-length-of-valid-subsequence-i.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums) {
        vector<int> parityList;
        for (int num : nums) {
            parityList.push_back(num % 2);
        }
        int sumOfOnes = accumulate(parityList.begin(), parityList.end(), 0);
        int alternatingLength = 1, previousParity = parityList[0];
        for (size_t i = 1; i < parityList.size(); ++i) {
            if (previousParity != parityList[i]) {
                alternatingLength += 1;
                previousParity = parityList[i];
            }
        }
        return max({sumOfOnes, (int)parityList.size() - sumOfOnes, alternatingLength});
    }
};
